src/ripple/nodestore/impl/DatabaseShardImp.h

```diff
@@ -146,11 +146,29 @@ class DatabaseShardImp : public DatabaseShard
         NodeObjectType type,
         Blob&& data,
         uint256 const& hash,
-        std::uint32_t seq) override;
+        std::uint32_t seq,
+        bool const etl) override;
+
+    void
+    sync() override{};
 
     std::shared_ptr<NodeObject>
     fetch(uint256 const& hash, std::uint32_t seq) override;
 
+    std::vector<std::shared_ptr<NodeObject>>
+    fetchBatch(std::vector<uint256> const& hashes) override
+    {
```

Comment:
@seelabs I agree that this is pretty ugly as is. The reason `fetchBatch()` is in the interface of `Database` is because `Database` is the only class accessible from the places that use `fetchBatch()`, usually via `app.getNodeStore()`. This function can't really be implemented by `DatabaseShard`, because we need the ledger sequence of each object in question in order to know what shard to fetch from. We can get that information if we want, but that would require some amount of massaging, and the code will never actually be used (only reporting calls `fetchBatch()`, and reporting doesn't use shards).

Really `fetchBatch()` was written for use with Cassandra, because we can leverage server side and cluster side parallelism. We also implemented it for nudb, since reporting does support nudb. `fetchBatch()` for nudb just calls `fetch()` in a loop, so it's not really a batch fetch; but implementing the function makes the code cleaner at the call site.

Here is what I propose we do:
1. Implement `fetchBatch()` for the remaining `Backend`s (`RocksDBFactory`, `MemoryFactory`, `NullFactory`). This is easy, because they can just call their own `fetch()` method in a loop.
2. Remove `fetchBatch()` from `Database`, `DatabaseShard` and `DatabaseRotating`. Reporting doesn't support shards, and reporting also doesn't support online delete. So `fetchBatch()` will only be a member of `DatabaseNodeImp`. (side note, I also need to modify the code so reporting doesn't use `DatabaseRotating`; right now it's possible for reporting to use `DatabaseRotating`, but no actual rotating happens)
3. Cast to `DatabaseNodeImp` at the call site, and then call `fetchBatch()`. Something like `dynamic_cast<DatabaseNodeImp>(app.getNodeStore()).fetch_batch(...)`

What do you think?

Reply:
I implemented this proposed solution in `6e6dad9ad66c57d21891590ff9606d844583df1f`. Let me know what you think